### PR TITLE
io_uring: mark as dev preview and add perf documentation

### PR DESCRIPTION
**Pull request by alindima**
## Description of Changes

- Marks the `Async` block device io_engine as developer preview
- Adds docs about io_uring performance considerations in Firecracker

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: docs/api_requests/block-io-engine.md [hunk 1]
```diff
@@ -51,6 +51,56 @@ If a block device is configured with the `Async` io_engine on a host kernel
...
+For use-cases where the lowest latency on the aforementioned operations is
+desired, it is recommended to use the `Sync` IO engine.
+
+### Block IOPS and efficiency
+
+The `Async` engine performance potential is showcased when the block device
+backing files are placed on a physical disk that supports efficient parallel
+execution of requests, like an NVME drive.
+
+The performance measurements f we've done were made on NVME drives, and we've
```
Comment (serban300) [inline docs/api_requests/block-io-engine.md:80]: Nit: There's an extra `f`: `measurements f we've`

File: docs/api_requests/block-io-engine.md [hunk 2]
```diff
@@ -51,6 +51,56 @@ If a block device is configured with the `Async` io_engine on a host kernel
...
+discovered that:
+
+For __read__ workloads which operate on data that is not present in the
+host page cache, the performance improvement for `Async` is about 1.5x-3x in
+overall efficiency (IOPS per CPU load) and up to 30x when disregarding the
```
Comment (serban300) [inline docs/api_requests/block-io-engine.md:85]: I would say `30x in throughput`

File: docs/api_requests/block-io-engine.md [hunk 3]
```diff
@@ -51,6 +51,56 @@ If a block device is configured with the `Async` io_engine on a host kernel
...
+This translates to higher latencies on either of these operations:
+
+- API call duration for block device config
+- Boot time for VMs started via JSON config files
+- Snapshot restore time
+
+For use-cases where the lowest latency on the aforementioned operations is
+desired, it is recommended to use the `Sync` IO engine.
+
+### Block IOPS and efficiency
```
Comment (serban300) [inline docs/api_requests/block-io-engine.md:74]: I would also mention that when having multiple VMs it's recommended to distribute the backing files equally among all the NVMEs available on the host for better performance.

File: docs/api_requests/block-io-engine.md (hunk 1 above)
Comment (alindima) [reply docs/api_requests/block-io-engine.md:80]: removed it

File: docs/api_requests/block-io-engine.md (hunk 2 above)
Comment (alindima) [reply docs/api_requests/block-io-engine.md:85]: said: `30x in total IOPS`

File: docs/api_requests/block-io-engine.md (hunk 3 above)
Comment (alindima) [reply docs/api_requests/block-io-engine.md:74]: done

File: docs/api_requests/block-io-engine.md [hunk 4]
```diff
@@ -49,10 +51,13 @@ If a block device is configured with the `Async` io_engine on a host kernel
 older than 5.10.51, the API call will return a 400 Bad Request, with a
 suggestive error message.
 
-## Recommendations
+## Performance considerations
 
-In order to get the higher disk IO throughput, it is recommended to use the
-`Async` io_engine on hosts that support it.
+An important thing to note is that the performance is strictly tied to the host
```
Comment (raduiliescu) [inline docs/api_requests/block-io-engine.md:56]: I would just say The performance is strictly.. Drop the first part which doesn't add up information.

File: docs/api_requests/block-io-engine.md [hunk 5]
```diff
@@ -65,3 +70,82 @@ This translates to higher latencies on either of these operations:
 
 For use-cases where the lowest latency on the aforementioned operations is
 desired, it is recommended to use the `Sync` IO engine.
+
+### Block IOPS and efficiency
+
```
Comment (raduiliescu) [inline docs/api_requests/block-io-engine.md:75]: Do we also mention the downsides like increased startup time, cpu usage and maybe memory, that might be seen in oversubscription enviroments.

**Review by raduiliescu — COMMENTED** [review]
Changes look good. See 2 minor comments.

File: docs/api_requests/block-io-engine.md (hunk 5 above)
Comment (alindima) [reply docs/api_requests/block-io-engine.md:75]: We do.
We talk about the increased startup time in the previous paragraph and about the increased CPU consumption in the paragraph about write workloads

File: docs/api_requests/block-io-engine.md (hunk 4 above)
Comment (alindima) [reply docs/api_requests/block-io-engine.md:56]: done

**Review by serban300 — APPROVED** [review]

**Review by raduiliescu — APPROVED** [review]

**Pull request merged**